File: rust/src/day4.rs
use crate::util::load;

type Pair = Vec<u32>;

fn load_pairs(filename: &str) -> Vec<Pair> {
    load::<String>(filename)
        .into_iter()
        .map(|l| {
            l.split(&['-', ','])
                .map(|n| n.parse::<u32>().unwrap())
                .collect()
        })
        .collect()
}

pub fn part1() -> usize {
    let pairs: Vec<Pair> = load_pairs("data/day4.txt");
    pairs
        .into_iter()
        .filter(|p| (p[0] <= p[2] && p[1] >= p[3]) || (p[0] >= p[2] && p[1] <= p[3]))
        .count()
}

pub fn part2() -> usize {
    let pairs: Vec<Pair> = load_pairs("data/day4.txt");
    pairs
        .into_iter()
        .filter(|p| {
            (p[2] <= p[0] && p[0] <= p[3])
                || (p[2] <= p[1] && p[1] <= p[3])
                || (p[0] <= p[2] && p[2] <= p[1])
                || (p[0] <= p[3] && p[3] <= p[1])
        })
        .count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let count = part1();
        println!("Fully contains count {}", count);
        assert_eq!(count, 562);
    }

    #[test]
    fn test_part2() {
        let count = part2();
        println!("Overlap count {}", count);
        assert_eq!(count, 924);
    }
}

File: python/day4.py
import re
from typing import Tuple

from util import load

SectionRange = Tuple[int, int]


def parse_line(line: str) -> Tuple[SectionRange, SectionRange]:
    num = [int(n) for n in re.split(r'[,-]', line)]
    return (num[0], num[1]), (num[2], num[3])


def full_overlap(sr1, sr2) -> bool:
    return (sr1[0] <= sr2[0] and sr2[1] <= sr1[1]) or (sr2[0] <= sr1[0] and sr1[1] <= sr2[1])


def any_overlap(sr1, sr2) -> bool:
    return (sr1[0] <= sr2[0] <= sr1[1]) or \
        (sr1[0] <= sr2[1] <= sr1[1]) or \
        (sr2[0] <= sr1[0] <= sr2[1]) or \
        (sr2[0] <= sr1[1] <= sr2[1])


def part1(pairs):
    return len([pair for pair in pairs if full_overlap(*pair)])


def part2(pairs):
    return len([pair for pair in pairs if any_overlap(*pair)])


def main():
    lines = load('day4.txt')
    pairs = [parse_line(line) for line in lines]

    count1 = part1(pairs)
    print(f'Part 1: {count1}')
    assert count1 == 562

    count2 = part2(pairs)
    print(f'Part 2: {count2}')
    assert count2 == 924


if __name__ == "__main__":
    main()
